plugins/outputs/postgresql/utils/utils.go

```diff
@@ -0,0 +1,203 @@
+	case uint64:
+		return PgNumeric
+	case int64, int, uint, uint32:
+		return PgBigInt
+	case int32:
+		return PgInteger
+	case int16, int8:
+		return PgSmallInt
+	case float64:
+		return PgDoublePrecision
+	case float32:
+		return PgReal
+	case string:
+		return PgText
+	case time.Time:
+		return PgTimestampWithTimeZone
+	default:
+		log.Printf("E! Unknown datatype %T(%v)", value, value)
+		return PgText
+	}
+}
+
+// PgTypeCanContain tells you if one PostgreSQL data type can contain
+// the values of another without data loss.
+func PgTypeCanContain(canThis PgDataType, containThis PgDataType) bool {
+	switch canThis {
+	case containThis:
+		return true
+	case PgBigInt:
+		return containThis == PgInteger || containThis == PgSmallInt
+	case PgInteger:
+		return containThis == PgSmallInt
+	case PgDoublePrecision, PgReal: // You can store a real in a double, you just lose precision
+		return containThis == PgReal || containThis == PgBigInt || containThis == PgInteger || containThis == PgSmallInt
+	case PgNumeric:
+		return containThis == PgBigInt || containThis == PgSmallInt || containThis == PgInteger || containThis == PgReal || containThis == PgDoublePrecision
+	case PgTimestampWithTimeZone:
+		return containThis == PgTimestampWithoutTimeZone
+	default:
+		return false
+	}
+}
+
+// pgxLogger makes telegraf.Logger compatible with pgx.Logger
+type PGXLogger struct {
+	telegraf.Logger
+}
+
+func (l PGXLogger) Log(_ context.Context, level pgx.LogLevel, msg string, data map[string]interface{}) {
+	switch level {
+	case pgx.LogLevelError:
+		l.Errorf("PG %s - %+v", msg, data)
+	case pgx.LogLevelWarn:
+		l.Warnf("PG %s - %+v", msg, data)
+	case pgx.LogLevelInfo, pgx.LogLevelNone:
+		l.Infof("PG %s - %+v", msg, data)
+	case pgx.LogLevelDebug, pgx.LogLevelTrace:
+		l.Debugf("PG %s - %+v", msg, data)
+	default:
+		l.Debugf("PG %s - %+v", msg, data)
+	}
+}
+
+// GenerateInsert returns a SQL statement to insert values in a table
+// with $X placeholders for the values
+func GenerateInsert(fullSanitizedTableName string, columns []string) string {
+	valuePlaceholders := make([]string, len(columns))
+	quotedColumns := make([]string, len(columns))
+	for i, column := range columns {
+		valuePlaceholders[i] = fmt.Sprintf("$%d", i+1)
+		quotedColumns[i] = QuoteIdent(column)
+	}
+
+	columnNames := strings.Join(quotedColumns, ",")
+	values := strings.Join(valuePlaceholders, ",")
+	return fmt.Sprintf(insertIntoSQLTemplate, fullSanitizedTableName, columnNames, values)
+}
+
+func GetTagID(metric telegraf.Metric) int64 {
+	hash := fnv.New64a()
+	for _, tag := range metric.TagList() {
+		_, _ = hash.Write([]byte(tag.Key))
+		_, _ = hash.Write([]byte{0})
+		_, _ = hash.Write([]byte(tag.Value))
+		_, _ = hash.Write([]byte{0})
+	}
+	// Convert to int64 as postgres does not support uint64
+	return int64(hash.Sum64())
+}
+
+// WaitGroup is similar to sync.WaitGroup, but allows interruptable waiting (e.g. a timeout).
+type WaitGroup struct {
+	count int32
+	done chan struct{}
+}
```

Comment:
I don't understand the need of this. All functions in the wait group receive a context that can be canceled, why do you need this instead of doing
```golang
p.dbContextCancel()
wg.Wait()
```
with a "ususal" wait-group?

Reply:
I'm really not sure what you're referring to. None of the functions in a WaitGroup take a context. https://golang.org/pkg/sync/#WaitGroup

Reply:
Well you are using your wait-group implementation in one spot with
```golang
		p.writeWaitGroup = utils.NewWaitGroup()
		for i := 0; i < maxConns; i++ {
			p.writeWaitGroup.Add(1)
			go p.writeWorker(p.dbContext)
		}
```
so the function in the wait-group do have a cancel-able context doesn't it? So you can cancel `writeWorker` and just wait until all of them are finished. So why do you need to be able to cancel the wait-group? Or am I missing the intention behind your implementation?

Reply:
Ah, ok, I get what you're saying now.  
The problem is that we don't want to cancel. The wait group is used when gracefully shutting down the plugin. We can't just cancel the context, as that'll result in data loss. We stop sending data to it, and then give it a short amount of time to finish processing the current data. So we need to block and continue when one of 2 events happen: Either all the goroutines complete, or the timeout fires. When the timeout fires, then we cancel the context.